File: toggly-docusaurus-edge-sdk/libs/docusaurus-plugin/smoke-ws.mjs
import WebSocket from 'ws';

const appKey = process.env.TOGGLY_SMOKE_APP_KEY_FRONTEND;

if (!appKey) {
  process.exit(0);
}

const ws = new WebSocket(`wss://definitions.toggly.io/${appKey}/ws`);

const timeout = setTimeout(() => {
  ws.close();
  throw new Error('WebSocket timed out after 10 seconds');
}, 10_000);

ws.on('message', (data) => {
  const parsed = JSON.parse(data.toString());
  if (parsed.type === 'ping') return; // skip ping messages
  clearTimeout(timeout);
  if (!parsed.type) {
    throw new Error(`Missing type field in message: ${data}`);
  }
  if (!['sync', 'definitions', 'evaluated'].includes(parsed.type)) {
    throw new Error(`Expected type=sync, definitions, or evaluated, received ${parsed.type}`);
  }
  if (parsed.type === 'sync') {
    if (!parsed.etag) {
      throw new Error('Missing etag in sync message');
    }
    if (!parsed.lastUpdated) {
      throw new Error('Missing lastUpdated in sync message');
    }
  } else if (!parsed.timestamp) {
    throw new Error('Missing timestamp in message');
  }
  ws.close();
});

ws.on('error', (err) => {
  clearTimeout(timeout);
  throw err;
});

File: toggly-rust/toggly/tests/smoke_ws.rs
use futures_util::StreamExt;
use tokio_tungstenite::{connect_async, tungstenite::Message};

#[tokio::test]
async fn smoke_websocket_connection() {
    let app_key = match std::env::var("TOGGLY_SMOKE_APP_KEY_BACKEND") {
        Ok(v) if !v.is_empty() => v,
        _ => return,
    };

    let url = format!("wss://definitions.toggly.io/{}/ws", app_key);
    let (mut ws_stream, _) =
        tokio::time::timeout(std::time::Duration::from_secs(10), connect_async(&url))
            .await
            .expect("WebSocket connect timed out")
            .expect("WebSocket connect failed");

    let deadline = tokio::time::Instant::now() + std::time::Duration::from_secs(15);
    loop {
        let msg = tokio::time::timeout_at(deadline, ws_stream.next())
            .await
            .expect("WebSocket read timed out")
            .expect("WebSocket stream ended")
            .expect("WebSocket read error");

        if let Message::Text(text) = msg {
            let parsed: serde_json::Value = serde_json::from_str(&text).expect("Invalid JSON");
            if parsed["type"] == "ping" {
                continue;
            }
            assert!(
                parsed["type"] == "definitions" || parsed["type"] == "evaluated",
                "Expected type=definitions or evaluated, got {}",
                parsed["type"]
            );
            assert!(parsed.get("timestamp").is_some(), "Missing timestamp field");
            break;
        } else {
            panic!("Expected text message, got {:?}", msg);
        }
    }
}
